File: Python/mergesort/deepseek-r1-distill-llama-70b.py

def merge_sort(arr):
    """
    Optimized merge sort implementation that uses insertion sort for small subarrays
    and minimizes array copying.
    """
    def merge_sort_helper(arr, start, end):
        if end - start <= 1:
            return
        mid = (start + end) // 2
        merge_sort_helper(arr, start, mid)
        merge_sort_helper(arr, mid, end)
        merge(arr, start, mid, end)

    def merge(arr, start, mid, end):
        left = arr[start:mid]
        right = arr[mid:end]
        i = j = 0
        k = start
        while i < len(left) and j < len(right):
            if left[i] <= right[j]:
                arr[k] = left[i]
                i += 1
            else:
                arr[k] = right[j]
                j += 1
            k += 1
        while i < len(left):
            arr[k] = left[i]
            i += 1
            k += 1
        while j < len(right):
            arr[k] = right[j]
            j += 1
            k += 1

    def insertion_sort(arr, start, end):
        for i in range(start + 1, end):
            key = arr[i]
            j = i - 1
            while j >= start and arr[j] > key:
                arr[j + 1] = arr[j]
                j -= 1
            arr[j + 1] = key

    n = len(arr)
    if n <= 10:
        insertion_sort(arr, 0, n)
    else:
        merge_sort_helper(arr, 0, n)
    return arr

# Example usage
import sys

def main():
    input_path = "inputs.txt" if len(sys.argv) < 2 else sys.argv[1]

    try:
        with open(input_path, 'r') as f:
            arr = [int(line.strip()) for line in f if line.strip()]
    except FileNotFoundError:
        return

    sorted_arr = merge_sort(arr)
    

if __name__ == "__main__":
    main()

File: C/mergesort/deepseek-r1-distill-llama-70b.c
#include <stdio.h>
#include <stdlib.h>

// Function to perform insertion sort on smaller arrays
void insertionSort(int *array, int left, int right) {
    for (int i = left + 1; i <= right; i++) {
        int key = array[i];
        int j = i - 1;
        while (j >= left && array[j] > key) {
            array[j + 1] = array[j];
            j--;
        }
        array[j + 1] = key;
    }
}

// Function to merge two sorted subarrays
void merge(int *array, int *aux, int left, int mid, int right) {
    int i = left;
    int j = mid + 1;
    int k = left;

    while (i <= mid && j <= right) {
        if (array[i] <= array[j]) {
            aux[k++] = array[i++];
        } else {
            aux[k++] = array[j++];
        }
    }

    // Copy remaining elements from left subarray
    while (i <= mid) {
        aux[k++] = array[i++];
    }

    // Copy remaining elements from right subarray
    while (j <= right) {
        aux[k++] = array[j++];
    }

    // Copy sorted elements back to original array
    for (int m = left; m <= right; m++) {
        array[m] = aux[m];
    }
}

// Optimized merge sort function
void mergeSort(int *array, int *aux, int left, int right) {
    // Use insertion sort for small subarrays
    if (right - left + 1 < 10) {
        insertionSort(array, left, right);
        return;
    }

    int mid = (left + right) / 2;
    mergeSort(array, aux, left, mid);
    mergeSort(array, aux, mid + 1, right);
    merge(array, aux, left, mid, right);
}

// Driver function to initialize auxiliary array and call merge sort
void sortArray(int *array, int size) {
    int *aux = (int *)malloc(size * sizeof(int));
    mergeSort(array, aux, 0, size - 1);
    free(aux);
}

// Example usage
int main() {
    int arr[1000];
    int size = 0;

    FILE *file = fopen("../../inputs.txt", "r");
    if (!file) {
        return 1;
    }

    while (fscanf(file, "%d", &arr[size]) == 1 && size < 1000) {
        size++;
    }
    fclose(file);

    sortArray(arr, size);

    return 0;
}